xml_converter/src/xml_converter.cpp

```diff
@@ -61,30 +63,115 @@ void write_xml_file(string xml_filepath, map<string, Category>* marker_categorie
     outfile.close();
 }
 
+void create_proto_categories(waypoint::Category proto_category, set<string>* proto_categories) {
+    proto_categories->insert(proto_category.name());
+    for (int i = 0; i < proto_category.children_size(); i++) {
+        create_proto_categories(proto_category.children(i), proto_categories);
+    }
+}
+
+void remove_proto_child(waypoint::Category* proto_category, set<string> category_includes) {
+    int keep = 0;
+    for (int i = 0; i < proto_category->children_size(); i++) {
+        vector<string> split_categories = split(proto_category->children(i).name(), ".");
+        auto pos = category_includes.find(split_categories[split_categories.size() - 1]);
+        if (pos != category_includes.end()) {
+            if (keep < i) {
+                proto_category->mutable_children()->SwapElements(i, keep);
+                if (proto_category->children(i).children_size() >= 0) {
+                    for (int j = 0; j < proto_category->children_size(); j++) {
+                        remove_proto_child(proto_category->mutable_children(j), category_includes);
+                    }
+                }
+            }
+            ++keep;
+        }
+    }
+
+    proto_category->mutable_children()->DeleteSubrange(keep, proto_category->children_size() - keep);
+}
+
 void write_protobuf_file(string proto_filepath, map<string, Category>* marker_categories, vector<Parseable*>* parsed_pois) {
-    ofstream outfile;
-    waypoint::Waypoint proto_message;
+    waypoint::Waypoint proto_pois;
+    waypoint::Waypoint all_categories;
+    waypoint::Waypoint output_message;
+    std::set<int> map_ids;
+    std::set<string> proto_categories;
+    std::set<string> category_includes;
```

Comment:
I am not a fan of defining all the variables at the start of a function, it makes it difficult to see where they are used. Very old versions of C require this. These should be moved to more logical locations such as before whatever blocks need them, or if they don't need to be scoped to the function: within whatever block scope makes sense.

Of course there are times where the most logical place for some variables to be defined is indeed the top of the  function, but this is not one of them.

Reply:
That makes logical sense to me. Declare when they need to be used.